File: sdks/typescript/src/agent/protocol.ts
/**
 * Frames exchanged over the agent socket; shapes match ADR-128.
 * @see dev/docs/adr/128-connected-agents.md
 */

export const PROTOCOL_VERSION = 1;

/** One conversation message, OpenAI style. Extra keys are carried as is. */
export interface AgentMessage {
  role: string;
  content?: unknown;
  [key: string]: unknown;
}

/** The value of one run parameter as the platform sends it. */
export type AgentParameterValue = string | number | boolean;

/** A JSON Schema object as the SDK sends it in `register`. */
export type JsonSchemaObject = Record<string, unknown>;

export interface RegisterSdk {
  name: string;
  version: string;
  language: string;
}

export interface RegisterInstance {
  id: string;
  hostname: string;
  username: string;
  pid: number;
  startedAt: string;
  label?: string;
  inFlightCallIds: string[];
}

export interface RegisterAgent {
  name: string;
  environment: string;
  parameters: JsonSchemaObject;
  concurrency?: number;
  timeoutMs?: number;
  sticky?: boolean;
}

export interface RegisterFrame {
  type: "register";
  protocol: typeof PROTOCOL_VERSION;
  sdk: RegisterSdk;
  instance: RegisterInstance;
  agents: RegisterAgent[];
}

export interface AckFrame {
  type: "ack";
  protocol: typeof PROTOCOL_VERSION;
  callId: string;
}

export interface CallError {
  code: string;
  message: string;
}

export type ResultFrame =
  | {
      type: "result";
      protocol: typeof PROTOCOL_VERSION;
      callId: string;
      output: unknown;
      session?: unknown;
    }
  | {
      type: "result";
      protocol: typeof PROTOCOL_VERSION;
      callId: string;
      error: CallError;
    };

export interface DeregisterFrame {
  type: "deregister";
  protocol: typeof PROTOCOL_VERSION;
}

/** Everything the SDK sends. */
export type ClientFrame = RegisterFrame | AckFrame | ResultFrame | DeregisterFrame;

/**
 * Who can target the agent: everyone, the owner of the registering key, or
 * the project through this machine.
 */
export type RegisteredAgentScope =
  | { kind: "shared" }
  | { kind: "owner" }
  | { kind: "host"; hostLabel: string };

export interface RegisteredAgent {
  name: string;
  environment: string;
  id: string;
  url: string;
  parameterNotes: string[];
  scope: RegisteredAgentScope;
}

export interface RegisteredFrame {
  type: "registered";
  protocol: number;
  agents: RegisteredAgent[];
  heartbeatIntervalMs: number;
  instanceId: string;
}

export interface RefusedFrame {
  type: "refused";
  protocol: number;
  code: string;
  message: string;
  /** Extra data for one code, for example the projects a key reaches under `project_required`. */
  meta?: Record<string, unknown>;
}

export interface CallRun {
  scenarioRunId?: string;
  scenarioName?: string;
  batchRunId?: string;
}

export interface CallFrame {
  type: "call";
  protocol: number;
  callId: string;
  agentId: string;
  threadId: string;
  messages: AgentMessage[];
  newMessages: AgentMessage[];
  params: Record<string, AgentParameterValue>;
  session: unknown;
  traceparent: string | null;
  /** Epoch milliseconds, null when the platform sent none. */
  deadlineAt: number | null;
  run: CallRun;
}

export interface CancelFrame {
  type: "cancel";
  protocol: number;
  callId: string;
}

/** Everything the platform sends. */
export type ServerFrame = RegisteredFrame | RefusedFrame | CallFrame | CancelFrame;

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const isString = (value: unknown): value is string => typeof value === "string";

const isMessageList = (value: unknown): value is AgentMessage[] =>
  Array.isArray(value) && value.every((item) => isRecord(item) && isString(item.role));

const isStringList = (value: unknown): value is string[] =>
  Array.isArray(value) && value.every(isString);

/** A platform that predates the scope field registers everything as shared. */
const readScope = (value: unknown): RegisteredAgentScope => {
  if (!isRecord(value)) return { kind: "shared" };
  if (value.kind === "owner") return { kind: "owner" };
  if (value.kind === "host" && isString(value.hostLabel))
    return { kind: "host", hostLabel: value.hostLabel };
  return { kind: "shared" };
};

/** One agent of a registered frame, or null when the entry is malformed. */
const readRegisteredAgent = (entry: unknown): RegisteredAgent | null => {
  if (!isRecord(entry)) return null;
  if (!isString(entry.name)) return null;
  let id: string | null = null;
  if (isString(entry.id)) {
    id = entry.id;
  } else if (isString(entry.agentId)) {
    id = entry.agentId;
  }
  if (id === null) return null;
  return {
    name: entry.name,
    environment: isString(entry.environment) ? entry.environment : "",
    id,
    url: isString(entry.url) ? entry.url : "",
    parameterNotes: isStringList(entry.parameterNotes) ? entry.parameterNotes : [],
    scope: readScope(entry.scope),
  };
};

const readRegistered = (frame: Record<string, unknown>): RegisteredFrame | null => {
  if (!Array.isArray(frame.agents)) return null;
  if (!isString(frame.instanceId)) return null;
  const agents: RegisteredAgent[] = [];
  for (const entry of frame.agents) {
    const agent = readRegisteredAgent(entry);
    if (agent === null) return null;
    agents.push(agent);
  }
  return {
    type: "registered",
    protocol: typeof frame.protocol === "number" ? frame.protocol : PROTOCOL_VERSION,
    agents,
    heartbeatIntervalMs:
      typeof frame.heartbeatIntervalMs === "number" ? frame.heartbeatIntervalMs : 10_000,
    instanceId: frame.instanceId,
  };
};

const readRefused = (frame: Record<string, unknown>): RefusedFrame => ({
  type: "refused",
  protocol: typeof frame.protocol === "number" ? frame.protocol : PROTOCOL_VERSION,
  code: isString(frame.code) ? frame.code : "agent_register_refused",
  message: isString(frame.message) ? frame.message : "The platform refused the registration.",
  ...(isRecord(frame.meta) ? { meta: frame.meta } : {}),
});

const readParams = (value: unknown): Record<string, AgentParameterValue> => {
  if (!isRecord(value)) return {};
  const params: Record<string, AgentParameterValue> = {};
  for (const [name, item] of Object.entries(value)) {
    if (typeof item === "string" || typeof item === "number" || typeof item === "boolean") {
      params[name] = item;
    }
  }
  return params;
};

/** A deadline as epoch milliseconds, from a number or an ISO string. */
const readDeadline = (value: unknown): number | null => {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (isString(value)) {
    const parsed = Date.parse(value);
    return Number.isNaN(parsed) ? null : parsed;
  }
  return null;
};

const readCall = (frame: Record<string, unknown>): CallFrame | null => {
  if (!isString(frame.callId)) return null;
  if (!isString(frame.agentId)) return null;
  const messages = isMessageList(frame.messages) ? frame.messages : [];
  const run = isRecord(frame.run) ? frame.run : {};
  return {
    type: "call",
    protocol: typeof frame.protocol === "number" ? frame.protocol : PROTOCOL_VERSION,
    callId: frame.callId,
    agentId: frame.agentId,
    threadId: isString(frame.threadId) ? frame.threadId : "",
    messages,
    newMessages: isMessageList(frame.newMessages) ? frame.newMessages : messages,
    params: readParams(frame.params),
    session: frame.session === undefined ? null : frame.session,
    traceparent: isString(frame.traceparent) ? frame.traceparent : null,
    deadlineAt: readDeadline(frame.deadlineAt),
    run: {
      ...(isString(run.scenarioRunId) ? { scenarioRunId: run.scenarioRunId } : {}),
      ...(isString(run.scenarioName) ? { scenarioName: run.scenarioName } : {}),
      ...(isString(run.batchRunId) ? { batchRunId: run.batchRunId } : {}),
    },
  };
};

/**
 * Reads one text message into a typed frame, or null when this protocol
 * version doesn't know it. Unknown/malformed frames are dropped, not thrown,
 * so a newer platform never crashes an older SDK.
 */
export function parseServerFrame(raw: string): ServerFrame | null {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    return null;
  }
  if (!isRecord(parsed)) return null;
  if (!isString(parsed.type)) return null;
  switch (parsed.type) {
    case "registered":
      return readRegistered(parsed);
    case "refused":
      return readRefused(parsed);
    case "call":
      return readCall(parsed);
    case "cancel":
      return isString(parsed.callId)
        ? {
            type: "cancel",
            protocol: typeof parsed.protocol === "number" ? parsed.protocol : PROTOCOL_VERSION,
            callId: parsed.callId,
          }
        : null;
    default:
      return null;
  }
}

/** One frame as the text the socket carries. */
export function serializeFrame(frame: ClientFrame): string {
  return JSON.stringify(frame);
}

/** The trace id of a W3C traceparent header, or null when it does not parse. */
export function traceIdFromTraceparent(traceparent: string | null | undefined): string | null {
  if (!traceparent) return null;
  const parts = traceparent.trim().split("-");
  const traceId = parts[1];
  if (parts.length < 4 || !traceId || !/^[0-9a-f]{32}$/i.test(traceId)) return null;
  return traceId.toLowerCase();
}

File: sdks/python/src/langwatch/agent/protocol.py
"""Frames of the connected agent protocol.

Every frame is one JSON text message over the socket and carries `type` and
`protocol`. The shapes match the contract table of ADR-128. Builders return
plain dicts so the client serializes them with `json.dumps` and nothing else.
"""

from __future__ import annotations

from typing import Any, Literal

from typing_extensions import NotRequired, TypedDict

PROTOCOL_VERSION = 1

CONNECT_PATH = "/api/v1/agents/connect"

# Frame sizes in bytes. The socket frame cap follows the platform default; a
# self-hosted deployment can raise it, and the client only needs to accept
# what the platform sends.
MAX_FRAME_BYTES = 64 * 1024 * 1024

Message = dict[str, Any]
"""One OpenAI-style chat message: a dict with `role` and `content`."""

ErrorCode = Literal[
    "agent_busy",
    "agent_call_failed",
    "agent_call_timeout",
    "agent_parameter_invalid",
]


class SdkInfo(TypedDict):
    name: str
    version: str
    language: str


class InstanceInfo(TypedDict):
    id: str
    hostname: str
    username: str
    pid: int
    startedAt: str
    label: NotRequired[str]
    inFlightCallIds: list[str]


class AgentRegistration(TypedDict):
    name: str
    environment: str
    parameters: dict[str, Any]
    concurrency: NotRequired[int]
    timeoutMs: NotRequired[int]
    sticky: NotRequired[bool]


class RegisterFrame(TypedDict):
    type: Literal["register"]
    protocol: int
    sdk: SdkInfo
    instance: InstanceInfo
    agents: list[AgentRegistration]


class RegisteredAgent(TypedDict):
    name: str
    environment: str
    id: str
    url: str
    parameterNotes: list[str]


class RegisteredFrame(TypedDict):
    type: Literal["registered"]
    protocol: int
    agents: list[RegisteredAgent]
    heartbeatIntervalMs: int
    instanceId: str


class RefusedFrame(TypedDict):
    type: Literal["refused"]
    protocol: int
    code: str
    message: str


class CallRun(TypedDict, total=False):
    scenarioRunId: str
    scenarioName: str
    batchRunId: str


class CallFrame(TypedDict):
    type: Literal["call"]
    protocol: int
    callId: str
    agentId: str
    threadId: str
    messages: list[Message]
    newMessages: list[Message]
    params: dict[str, Any]
    session: Any
    traceparent: str | None
    deadlineAt: int | float | str
    run: CallRun


class AckFrame(TypedDict):
    type: Literal["ack"]
    protocol: int
    callId: str


class ResultError(TypedDict):
    code: str
    message: str


class ResultFrame(TypedDict):
    type: Literal["result"]
    protocol: int
    callId: str
    output: NotRequired[str | Message | list[Message]]
    session: NotRequired[Any]
    error: NotRequired[ResultError]


class CancelFrame(TypedDict):
    type: Literal["cancel"]
    protocol: int
    callId: str


class DeregisterFrame(TypedDict):
    type: Literal["deregister"]
    protocol: int


def register_frame(
    *,
    sdk: SdkInfo,
    instance: InstanceInfo,
    agents: list[AgentRegistration],
) -> RegisterFrame:
    return {
        "type": "register",
        "protocol": PROTOCOL_VERSION,
        "sdk": sdk,
        "instance": instance,
        "agents": agents,
    }


def ack_frame(call_id: str) -> AckFrame:
    return {"type": "ack", "protocol": PROTOCOL_VERSION, "callId": call_id}


def result_frame(
    *,
    call_id: str,
    output: str | Message | list[Message],
    session: Any = None,
) -> ResultFrame:
    frame: ResultFrame = {
        "type": "result",
        "protocol": PROTOCOL_VERSION,
        "callId": call_id,
        "output": output,
    }
    if session is not None:
        frame["session"] = session
    return frame


def error_result_frame(*, call_id: str, code: str, message: str) -> ResultFrame:
    return {
        "type": "result",
        "protocol": PROTOCOL_VERSION,
        "callId": call_id,
        "error": {"code": code, "message": message},
    }


def deregister_frame() -> DeregisterFrame:
    return {"type": "deregister", "protocol": PROTOCOL_VERSION}
